### PR TITLE
Terraform files to create AWS Codepipeline for Doorway CI/CD

### PR DESCRIPTION
**Pull request by glgh**
Tested on my personal AWS account but not yet on the BAHFA account

### CODE REVIEW

File: codepipeline/main.tf
Comment (chriscasto) [inline codepipeline/main.tf:11]: Just a quick note that this will need to be changed to the state bucket in doorway before merge

File: codepipeline/codepipeline.tf
Comment (chriscasto) [inline codepipeline/codepipeline.tf:191]: This enables read/write access to any S3 bucket from within this CodeBuild job, which is a security issue.  This needs to be scoped down to just the bucket(s) that it actually needs to read from and/or write to.

File: codepipeline/codepipeline.tf
Comment (chriscasto) [inline codepipeline/codepipeline.tf:204]: This is redundant with the more expansive statement above, but also provides access to any S3 bucket.

File: codepipeline/codepipeline.tf
Comment (chriscasto) [inline codepipeline/codepipeline.tf:197]: Why does CodeBuild need access to read bucket ACLs?

File: codepipeline/codepipeline.tf
Comment (chriscasto) [inline codepipeline/codepipeline.tf:198]: And why does it need to know what region the bucket is in?

File: codepipeline/codepipeline.tf
Comment (chriscasto) [inline codepipeline/codepipeline.tf:180]: This should be restricted to something like "arn:aws:logs:*:*:log-group:/aws/codebuild/*" at least, although PutLogEvents will need something more like "arn:aws:logs:*:*:log-group:/aws/codebuild/*:log-stream:*".

File: codepipeline/codepipeline.tf
Comment (chriscasto) [inline codepipeline/codepipeline.tf:222]: These need to be scoped to a specific report group.

File: codepipeline/codepipeline.tf
Comment (chriscasto) [inline codepipeline/codepipeline.tf:144]: This isn't a huge security risk since this policy should, in theory, only be used by CodePipeline, but it's generally a good idea to restrict access to only the resources that need to be accessed.  In this case, you would want to use the [arn for the CodeBuild project](https://docs.aws.amazon.com/service-authorization/latest/reference/list_awscodebuild.html#awscodebuild-project).

File: codepipeline/codepipeline.tf
Comment (chriscasto) [inline codepipeline/codepipeline.tf:78]: You will probably want to use [bucket_prefix](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket#bucket_prefix) instead of bucket to reduce the chance of collisions.

**Comment by glgh** [comment]
thanks for the first pass Chris! Let me work on these comments now.

File: codepipeline/main.tf
Comment (glgh) [reply codepipeline/main.tf]: Done.

File: codepipeline/codepipeline.tf
Comment (glgh) [reply codepipeline/codepipeline.tf]: Done.

File: codepipeline/codepipeline.tf
Comment (glgh) [reply codepipeline/codepipeline.tf]: Done.

File: codepipeline/codepipeline.tf
Comment (glgh) [reply codepipeline/codepipeline.tf]: Done.

File: codepipeline/codepipeline.tf
Comment (glgh) [reply codepipeline/codepipeline.tf]: Removed.

File: codepipeline/codepipeline.tf
Comment (glgh) [reply codepipeline/codepipeline.tf]: Removed.

File: codepipeline/codepipeline.tf
Comment (glgh) [reply codepipeline/codepipeline.tf]: Done.

File: codepipeline/codepipeline.tf
Comment (glgh) [reply codepipeline/codepipeline.tf]: Changed to the CodeBuild project arn

File: codepipeline/codepipeline.tf
Comment (glgh) [reply codepipeline/codepipeline.tf]: Done.

**Comment by glgh** [comment]
PTAL again! Addressed the comments and tested the pipeline again to make sure it still works on my personal.

File: codepipeline/vars.tf
Comment (chriscasto) [inline codepipeline/vars.tf:21]: This should have some validation on it

File: codepipeline/vars.tf
Comment (chriscasto) [inline codepipeline/vars.tf:26]: This should have validation too, and should probably default to "main"

**Review by chriscasto — CHANGES_REQUESTED** [review]
Just a couple small changes

**Comment by chriscasto** [comment]
Is the goal to keep this as a sort of sample configuration or is this ultimately going to morph into the actual set of templates for deploying the bloom code?  If the latter, it should be in a directory named something like "app-pipeline" to distinguish it from the infra pipeline that will be added later.

**Comment by glgh** [comment]
> Is the goal to keep this as a sort of sample configuration or is this ultimately going to morph into the actual set of templates for deploying the bloom code? If the latter, it should be in a directory named something like "app-pipeline" to distinguish it from the infra pipeline that will be added later.

Renamed.

**Review by chriscasto — APPROVED** [review]
I think we'll want to make those validation regexes a bit more specific in the future, but they should be okay for now as it's still a WIP.  Approved

**Pull request merged**
